flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java

```diff
@@ -59,31 +63,45 @@ public void configure(Configuration parameters) {
 
     /** Opens a Session to Cassandra . */
     @Override
-    public void open(int taskNumber, int numTasks) {
+    protected void postOpen() {
         this.session = cluster.connect();
-        super.open(taskNumber, numTasks);
     }
 
     /** Closes all resources used by Cassandra connection. */
     @Override
-    public void close() throws IOException {
+    protected void postClose() {
         try {
-            super.close();
-        } finally {
-            try {
-                if (session != null) {
-                    session.close();
-                }
-            } catch (Exception e) {
-                LOG.error("Error while closing session.", e);
+            if (session != null) {
+                session.close();
             }
-            try {
-                if (cluster != null) {
-                    cluster.close();
-                }
-            } catch (Exception e) {
-                LOG.error("Error while closing cluster.", e);
+        } catch (Exception e) {
+            LOG.error("Error while closing session.", e);
+        }
+        try {
+            if (cluster != null) {
+                cluster.close();
             }
+        } catch (Exception e) {
+            LOG.error("Error while closing cluster.", e);
         }
     }
+
+    protected static <T> CompletableFuture<T> listenableFutureToCompletableFuture(
+            final ListenableFuture<T> listenableFuture) {
+        CompletableFuture<T> completable = new CompletableFuture<T>();
+        Futures.addCallback(
+                listenableFuture,
+                new FutureCallback<T>() {
+                    @Override
+                    public void onSuccess(T result) {
+                        completable.complete(result);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable t) {
+                        completable.completeExceptionally(t);
+                    }
+                });
```

Comment:
this generates a new class on every call. Create a dedicated class that you pass the future to via the constructor.

Reply:
thanks for catching this. My bad ! How could I write this code that allocates with each send call ?!! :scream:

Reply:
My bad, thanks for pointing out.

Reply:
Fixed